src/mlpack/core/optimizers/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -0,0 +1,146 @@
+    SparseFunctionType& function,
+    arma::mat& iterate)
+{
+  double overallObjective = DBL_MAX;
+  double lastObjective;
+
+  // The order in which the functions will be visited.
+  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+      (function.NumFunctions() - 1), function.NumFunctions());
+
+  // Iterate till the objective is within tolerance or the maximum number of
+  // allowed iterations is reached. If maxIterations is 0, this will iterate
+  // till convergence.
+  for (size_t i = 1; i != maxIterations; ++i)
+  {
+    // Calculate the overall objective.
+    lastObjective = overallObjective;
+    overallObjective = 0;
+
+    for (size_t j = 0; j < function.NumFunctions(); ++j)
+    {
+      overallObjective += function.Evaluate(iterate, j);
+    }
+
+    // Output current objective function.
+    Log::Info << "Parallel SGD: iteration " << i << ", objective "
+      << overallObjective << "." << std::endl;
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Parallel SGD: converged to " << overallObjective
+        << "; terminating with failure. Try a smaller step size?"
+        << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
+        << "terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    // Shuffle for uniform sampling of functions by each thread.
+    visitationOrder = arma::shuffle(visitationOrder);
+
+    #pragma omp parallel
+    {
+      // Each processor gets a subset of the instances.
+      // Each subset is of size threadShareSize.
+      arma::Col<size_t> instances = ThreadShare(omp_get_thread_num(),
+          visitationOrder);
+      for (size_t j = 0; j < instances.n_elem; ++j)
+      {
+        // Each instance affects only some components of the decision variable.
+        // So the gradient is sparse.
+        arma::sp_mat gradient;
+
+        // Evaluate the sparse gradient.
+        function.Gradient(iterate, instances[j], gradient);
+
+        // Update the decision variable with non-zero components of the
+        // gradient.
+        for (size_t i = 0; i < gradient.n_cols; ++i)
+        {
+          // Iterate over the non-zero elements.
+          for (arma::sp_mat::iterator cur = gradient.begin_col(i);
+              cur != gradient.end_col(i); ++cur)
+          {
+            #pragma omp atomic
+            iterate(cur.row(), i) -= stepSize * (*cur);
+          }
+        }
+      }
+    }
+  }
+  Log::Info << "\n Parallel SGD terminated with objective : "
+    << overallObjective << std::endl;
+  return overallObjective;
+}
+
+template <typename DecayPolicyType>
+arma::Col<size_t> ParallelSGD<DecayPolicyType>::ThreadShare(
```

Comment:
The compiler might optimize this away, but in case if not, we could pass the thread share as a reference. Also do you think we should make shuffel optinal, that way we could entirely avoid the vector copy.